src/nunit.analyzers/ComparableTypes/ComparableTypesAnalyzer.cs

```diff
@@ -0,0 +1,125 @@
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using NUnit.Analyzers.Constants;
+using NUnit.Analyzers.Extensions;
+using NUnit.Analyzers.Helpers;
+using NUnit.Analyzers.Operations;
+
+namespace NUnit.Analyzers.ComparableTypes
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComparableTypesAnalyzer : BaseAssertionAnalyzer
+    {
+        private static readonly DiagnosticDescriptor descriptor = DiagnosticDescriptorCreator.Create(
+            id: AnalyzerIdentifiers.ComparibleTypes,
+            title: ComparableConstants.Title,
+            messageFormat: ComparableConstants.Message,
+            category: Categories.Assertion,
+            defaultSeverity: DiagnosticSeverity.Warning,
+            description: ComparableConstants.Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(descriptor);
+
+        private static readonly ImmutableHashSet<string> SupportedConstraints = ImmutableHashSet.Create(
+            NunitFrameworkConstants.NameOfIsLessThan,
+            NunitFrameworkConstants.NameOfIsLessThanOrEqualTo,
+            NunitFrameworkConstants.NameOfIsGreaterThan,
+            NunitFrameworkConstants.NameOfIsGreaterThanOrEqualTo);
+
+        protected override void AnalyzeAssertInvocation(OperationAnalysisContext context, IInvocationOperation assertOperation)
+        {
+            if (!AssertHelper.TryGetActualAndConstraintOperations(assertOperation,
+                out var actualOperation, out var constraintExpression))
+            {
+                return;
+            }
+
+            IMethodSymbol? constraintMethod = null;
+            IOperation? expectedOperation = null;
+
+            foreach (var constraintPartExpression in constraintExpression.ConstraintParts)
+            {
+                if (HasIncompatiblePrefixes(constraintPartExpression)
+                    || HasCustomComparer(constraintPartExpression)
+                    || constraintPartExpression.HasUnknownExpressions())
+                {
+                    return;
+                }
+
+                constraintMethod = constraintPartExpression.GetConstraintMethod();
+
+                if (constraintMethod == null)
+                    continue;
+
+                if (!SupportedConstraints.Contains(constraintMethod.Name))
+                    continue;
+
+                expectedOperation = constraintPartExpression.GetExpectedArgument();
+                break;
```

Comment:
Because I didn't know your example was possible, nor have I seen it before. This will also affect `SameAsOnValueTypesAnalyzer`. I will have to find a different way to deal with these.

Reply:
Implemented now.